portality/scripts/githubpri/github_prioritisation.py

```diff
@@ -0,0 +1,104 @@
+
+    print('[Start] update google sheet')
+
+    if gdrive_filename is None:
+        print('gdrive filename is not provided, skip updating google sheet')
+        sys.exit(0)
+    elif gdrive_key_path is None:
+        log.warning('gdrive json key path is not provided, skip updating google sheet')
+        sys.exit(1)
+
+    display_df = pd.concat({user: pri_df.drop('issue_url', axis=1)
+                            for user, pri_df in user_pri_map.items()},
+                           axis=1)
+    client = gdrive_sheet_serv.load_client(gdrive_key_path)
+    sh = client.open(gdrive_filename)
+
+    worksheet = create_or_load_worksheet(sh)
+
+    gdrive_sheet_serv.update_sheet_by_df(worksheet, display_df)
+
+    # assign title to issue_url's hyperlink
+    cells = []
+    for col_idx, (column_keys, titles) in enumerate(display_df.items()):
+        if 'title' not in column_keys:
+            continue
+        username, *_ = column_keys
+        titles = titles.dropna().apply(lambda x: x.replace('"', '""'))
+        cells = worksheet.range(3, col_idx + 1, len(titles) + 3, col_idx + 1)
+        for (row_idx, title), cell in zip(titles.items(), cells):
+            link = user_pri_map[username].loc[row_idx, 'issue_url']
+            cell.value = f'=HYPERLINK("{link}", "{title}")'
+        worksheet.update_cells(cells, ValueInputOption.user_entered)
+
+    gdrive_sheet_serv.apply_prilist_styles(worksheet, display_df)
+    print('[End] update google sheet')
+
+
+def main():
+    """
+    this script is to generate a excel for DOAJ github issue prioritisation queue for each user
+
+    you need github and google drive api key to run this script
+    DOAJ_GITHUB_KEY is github api key
+
+    DOAJ_PRILIST_KEY_PATH is json file path for google drive api
+    the DOAJ_PRILIST_KEY_PATH json file generated by console.cloud.google.com and should be enabled for
+    * 'google drive api'
+    * 'google sheet api'
+
+    """
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-u", "--username",
+                        help="github username")
+    parser.add_argument("-p", "--password",
+                        help="github password")
+    parser.add_argument("-r", "--rules",
+                        help="file path of rules excel", required=True)
+    parser.add_argument("-o", "--out",
+                        help="file path of local output excel")
+    parser.add_argument("-g", "--gdrive-name",
+                        help="excel name in google drive")
+
+    args = parser.parse_args()
+
+
+    priorities(args.rules,
+               outfile=args.out,
+               gdrive_filename=args.gdrive_name,
+               gdrive_key_path=os.environ.get('DOAJ_PRILIST_KEY_PATH'),
+               github_api_key=os.environ.get('DOAJ_GITHUB_KEY'),
+               github_username=args.username,
+               github_password=args.password,
+               )
```

Comment:
I'm not sure why some of the arguments are passed into this from the command line, and some are provided via the environment.  I'd prefer it if there was always a command line argument option, even if these can then be overridden or set by the environment.

Also, what's the difference between the `--password` argument (which I can pass my github key too) and the `DOAJ_GITHUB_KEY`?

Reply:
Because input password or API key in command line could have security issues.
* Shoulder Surfing -- anyone standing behind you or nearby surveillance cameras might be able to see and record it.
* lack password when screen sharing
* if other script need to use `github_prioritisation.py`, the password would be save to the script and that script could be uploaded to github when `git push`

we can use `dev.cfg` to save api key instead of `os.environ`, but `github_prioritisation.py` will have dependency to `from portality.core import app`

I don't know HTTPBasicAuth support API key before. I will update auth method